marshmallow_dataclass/__init__.py

```diff
@@ -271,7 +274,32 @@ def class_schema(
     Traceback (most recent call last):
     ...
     marshmallow.exceptions.ValidationError: {'name': ['Name too long']}
+    >>> @dataclasses.dataclass
+    ... class Simple:
+    ...     one: str = dataclasses.field()
+    ...     two: str = dataclasses.field()
+    >>> @dataclasses.dataclass
+    ... class ComplexNested:
+    ...     three: int = dataclasses.field()
+    ...     four: Simple = dataclasses.field()
+    >>> id(class_schema(ComplexNested)) == id(class_schema(ComplexNested))
+    True
+    >>> class_schema(ComplexNested) is class_schema(ComplexNested)
+    True
+    >>> id(class_schema(Simple)) == id(class_schema(Simple))
+    True
+    >>> class_schema(Simple) is class_schema(Simple)
+    True
+    >>> class_schema(Simple) is class_schema(ComplexNested)._declared_fields["four"].nested
+    True
+    >>> len(set([class_schema(ComplexNested), class_schema(ComplexNested, base_schema=None), class_schema(ComplexNested, None)]))
+    1
+    >>> len(set([class_schema(Simple), class_schema(Simple, base_schema=None), class_schema(Simple, None)]))
+    1
```

Comment:
Maybe this would be better in a separate test file than in the function's doc comment ?

Reply:
> I think it is acceptable in practice to return different instances when the function is invoked differently.

Yes, but in this instance it is unusual behavior and non-intuitive. E.g. giving the same dataclass and getting different schema instances back based on how you "called" the function.

> And if you really want to handle this case, I think you still shouldn't roll your own cache implementation. You can just create a proxy function without arguments on which you'll apply `lru_cache`.

It's a necessary case due to nested schemas. Two complex dataclasses that reference other custom dataclasses generate duplicates because the function to generate them gets called internally inside the marshmallow-dataclass functions. If I could have resolved this issue without making changes to the lib, I would have done so. If you have any suggestions as to how we can do that, that'll be awesome?

Reply:
@lovasoa Will move the tests to a separate file.